url9582d546a9_twentylemon_sudoku_generator_tgz-pJ8-sudoku_Top95J8/src/sudoku/SolverService.java

```diff
@@ -0,0 +1,110 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import sudoku.backtrack.BacktrackSolver;
+import sudoku.clp.ConstraintSolver;
+import sudoku.dlx.ExactCoverSolver;
+import sudoku.util.MapUtil;
+
+/**
+ * Service that runs all solvers each in their own thread. Once one thread
+ * finishes, all other threads are killed and the result is returned.
+ *
+ * @author Taras Mychaskiw
+ */
+public class SolverService {
+
+    private static SudokuBoard BOARD;   //board being solved
+    private static Class winner = null;
+    private static long time = 0;
+    private static long totalTime = 0;
+    private static final Map<Class,Integer> wins = new HashMap<>();
+    private static int totalRuns = 0;
+
+    /**
+     * list of all the solvers testing if the sudoku is well formed
+     */
+    private static final SudokuSolver BACK = new BacktrackSolver();
+    private static final SudokuSolver CLP = new ConstraintSolver();
+    private static final SudokuSolver DLX = new ExactCoverSolver();
+    private static final List<Callable<Integer>> CALLABLES = new LinkedList<Callable<Integer>>(){{
+        add(new Callable<Integer>(){
+            @Override
+            public Integer call(){
+                long time = System.currentTimeMillis();
+                Integer result = BACK.getFormity(BOARD);
+                setWinner(BacktrackSolver.class, System.currentTimeMillis() - time);
+                return result;
+            }
+        });
+        add(new Callable<Integer>(){
+            @Override
+            public Integer call(){
+                long time = System.currentTimeMillis();
+                Integer result = CLP.getFormity(BOARD);
+                setWinner(ConstraintSolver.class, System.currentTimeMillis() - time);
+                return result;
+            }
+        });
+        add(new Callable<Integer>(){
+            @Override
+            public Integer call(){
+                long time = System.currentTimeMillis();
+                Integer result = DLX.getFormity(BOARD);
+                setWinner(ExactCoverSolver.class, System.currentTimeMillis() - time);
+                return result;
+            }
+        });
+    }};
+    private static final int NUM_THREADS = CALLABLES.size();
+
+    /**
+     * Checks if the sudoku is well formed. Instead of picking a general
+     * solving strategy, a thread is opened for each strategy and they all test
+     * to see if the sudoku is well formed. After the first one returns,
+     * the other threads are killed.
+     *
+     * @param board the board to test
+     * @return the formity of the board
+     */
+    public static int getFormity(SudokuBoard board){
+        BOARD = board;
+        winner = null;
+        ExecutorService executor = Executors.newFixedThreadPool(NUM_THREADS);
+        int result = 0;
+        try {
+            result = executor.invokeAny(CALLABLES);
+        } catch (InterruptedException | ExecutionException ex){}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)